src/libstore/build/local-derivation-goal.cc

```diff
@@ -2138,12 +2138,29 @@ DrvOutputs LocalDerivationGoal::registerOutputs()
 
     /* The paths that can be referenced are the input closures, the
        output paths, and any paths that have been built via recursive
-       Nix calls. */
+       Nix calls; this list is restricted to the paths in the referenceablePaths
+       derivation attribute, if present. */
     StorePathSet referenceablePaths;
     for (auto & p : inputPaths) referenceablePaths.insert(p);
     for (auto & i : scratchOutputs) referenceablePaths.insert(i.second);
     for (auto & p : addedPaths) referenceablePaths.insert(p);
 
+    if (auto possibleReferences = parsedDrv->getStringsAttr("referenceablePaths")) {
+        StorePathSet spec;
+        for (auto & i : *possibleReferences) {
+            if (worker.store.isStorePath(i))
+                spec.insert(worker.store.parseStorePath(i));
+            else
+                throw BuildError("derivation contains an illegal possible reference '%s'", i);
+        }
+
+        StorePathSet tmp;
+        std::set_intersection(referenceablePaths.begin(), referenceablePaths.end(),
+                              spec.begin(), spec.end(),
+                              std::inserter(tmp, tmp.begin()));
```

Comment:
I think it would be useful to also log (`notice` I would say) the references that we erase here

Reply:
I'm a bit skeptical: I don't think it adds any information that the user doesn't already know, and I think there are usually too many paths in `referenceablePaths` for this to be useful and not spam.